## Q3. Confusion Matrix and Basic Metrics

---

### 🔍 এই Question-এ কী চাওয়া হয়েছে?

`true_label` ও `pred_label` ব্যবহার করে:
1. **TP, TN, FP, FN** manually হিসাব করতে হবে
2. **Accuracy, Precision, Recall** বের করতে হবে
3. Q3.3-এ 3-4 বাক্যে model-এর performance comment করতে হবে

### 🎯 এখান থেকে কী অর্জন করব?

শুধু accuracy দেখলেই model ভালো কিনা বোঝা যায় না — **precision ও recall** একসাথে দেখলে model-এর আসল দুর্বলতা বের করা যায়।

### 🧠 চারটি term মনে রাখার নিয়ম:

| | Predicted 1 | Predicted 0 |
|---|---|---|
| **Actual 1** | TP (সঠিকভাবে positive ধরেছে) | FN (positive মিস করেছে) |
| **Actual 0** | FP (ভুলভাবে positive বলেছে) | TN (সঠিকভাবে negative ধরেছে) |

| Metric | Formula | অর্থ |
|---|---|---|
| **Accuracy** | (TP+TN) / Total | মোট কতটি সঠিকভাবে predict হয়েছে |
| **Precision** | TP / (TP+FP) | যখন model "positive" বলে, কতটা নির্ভরযোগ্য |
| **Recall** | TP / (TP+FN) | আসল positive-গুলোর মধ্যে কতটা ধরতে পেরেছে |

### এই dataset-এর actual values:
- TP=**28**, TN=**27**, FP=**21**, FN=**24**
- Accuracy=**0.55**, Precision=**0.5714**, Recall=**0.5385**


### Q3.1 — TP, TN, FP, FN হিসাব করা

In [1]:
import pandas as pd
import numpy as np

# Now we'll load the dataset and define column names
df = pd.read_csv("02. Provided - module123_data.csv")
true_col = "true_label"
pred_col = "pred_label"
# Here we've loaded the data and set the column references.

# Now we'll compute TP, TN, FP, FN using boolean conditions
tp = ((df[true_col] == 1) & (df[pred_col] == 1)).sum()
tn = ((df[true_col] == 0) & (df[pred_col] == 0)).sum()
fp = ((df[true_col] == 0) & (df[pred_col] == 1)).sum()
fn = ((df[true_col] == 1) & (df[pred_col] == 0)).sum()
# Here we've calculated all four confusion matrix values.

print("TP:", tp)
print("TN:", tn)
print("FP:", fp)
print("FN:", fn)


TP: 28
TN: 27
FP: 21
FN: 24


`(df[true_col] == 1) & (df[pred_col] == 1)` → দুটো condition একসাথে True হলে TP।
`&` → element-wise AND operator — উভয় condition সত্য হলেই True।
`.sum()` → True গুলো গণনা করে।

**Result:** TP=28, TN=27, FP=21, FN=24


### Q3.2 — Accuracy, Precision, Recall হিসাব করা

In [2]:
# Now we'll compute accuracy, precision, and recall from TP/TN/FP/FN
accuracy  = (tp + tn) / (tp + tn + fp + fn) if (tp + tn + fp + fn) > 0 else 0
precision = tp / (tp + fp) if (tp + fp) > 0 else 0
recall    = tp / (tp + fn) if (tp + fn) > 0 else 0
# Here we've used the standard formulas with a safety check to avoid division by zero.

print("Accuracy: ", round(accuracy,  4))
print("Precision:", round(precision, 4))
print("Recall:   ", round(recall,    4))


Accuracy:  0.55
Precision: 0.5714
Recall:    0.5385


`if (tp + fp) > 0 else 0` → division by zero এড়ানোর জন্য safety check।
**Result:**
- Accuracy = **0.55** → 100টির মধ্যে 55টি সঠিকভাবে predict হয়েছে
- Precision = **0.5714** → model যখন "positive" বলে, 57% সময় সঠিক
- Recall = **0.5385** → আসল positive-গুলোর 54% model ধরতে পেরেছে


### Q3.3 — Short Answer [10 Marks]

> **Question:** In 3 to 4 sentences, briefly comment on the model using these three metrics. For example, is the model catching most positives (high recall) or being careful when it predicts positive (high precision)?

---

**Answer:**

> The model achieves an accuracy of 0.55, meaning it correctly classifies only 55 out of 100 samples, which is barely better than random guessing for a balanced dataset. The precision of 0.57 indicates that when the model predicts a positive label, it is correct about 57% of the time, suggesting a moderate but not reliable positive prediction. The recall of 0.54 shows that the model only identifies about 54% of the actual positive cases, meaning it misses nearly half of the true positives with 24 false negatives. Overall, all three metrics are close to 0.5, which strongly suggests that this model is performing poorly and requires significant improvement — possibly through better feature selection, more training data, or a more suitable algorithm.

> মডেলটির accuracy হলো ০.৫৫, অর্থাৎ এটি ১০০টি sample-এর মধ্যে মাত্র ৫৫টি সঠিকভাবে classify করতে পারে, যা একটি balanced dataset-এ random guessing-এর চেয়ে সামান্যই ভালো। Precision ০.৫৭ মানে হলো মডেলটি যখন কোনো positive label predict করে, তখন সে প্রায় ৫৭% ক্ষেত্রে সঠিক থাকে — যা মোটামুটি, তবে নির্ভরযোগ্য positive prediction বলা যায় না। Recall ০.৫৪ দেখাচ্ছে যে মডেলটি মাত্র ৫৪% actual positive case সনাক্ত করতে পারছে, অর্থাৎ ২৪টি false negative-সহ প্রকৃত positive-এর প্রায় অর্ধেকই তার নজর এড়িয়ে যাচ্ছে। সামগ্রিকভাবে, তিনটি metric-ই ০.৫-এর কাছাকাছি, যা স্পষ্টভাবে ইঙ্গিত দেয় যে মডেলটি দুর্বলভাবে কাজ করছে এবং এতে উল্লেখযোগ্য উন্নতি প্রয়োজন — সম্ভবত আরও ভালো feature selection, বেশি training data, অথবা আরও উপযুক্ত algorithm-এর মাধ্যমে।


---